src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs

```diff
@@ -53,29 +56,37 @@ public abstract class AbstractNavigateToTests
         internal readonly static PatternMatch s_emptyCamelCaseNonContiguousSubstringPatternMatch_NotCaseSensitive = new PatternMatch(PatternMatchKind.CamelCaseNonContiguousSubstring, true, false, ImmutableArray<Span>.Empty);
         internal readonly static PatternMatch s_emptyFuzzyPatternMatch_NotCaseSensitive = new PatternMatch(PatternMatchKind.Fuzzy, true, false, ImmutableArray<Span>.Empty);
 
+        public static IEnumerable<object[]> TrackingServiceTypes
+        {
+            get
+            {
+                yield return new object[] { null };
+                yield return new object[] { typeof(FirstDocIsActiveDocumentTrackingService.Factory) };
+                yield return new object[] { typeof(FirstDocIsVisibleDocumentTrackingService.Factory) };
+                yield return new object[] { typeof(FirstDocIsActiveAndVisibleDocumentTrackingService.Factory) };
+            }
+        }
+
         protected abstract TestWorkspace CreateWorkspace(string content, ExportProvider exportProvider);
         protected abstract string Language { get; }
 
-        protected async Task TestAsync(string content, Func<TestWorkspace, Task> body)
+        protected async Task TestAsync(string content, Func<TestWorkspace, Task> body, Type trackingServiceType)
         {
-            await TestAsync(content, body, outOfProcess: true);
-            await TestAsync(content, body, outOfProcess: false);
-        }
+            var catalog = TestExportProvider.EntireAssemblyCatalogWithCSharpAndVisualBasic;
+            if (trackingServiceType != null)
+            {
+                catalog = catalog.WithPart(trackingServiceType);
+            }
 
-        internal async Task TestAsync(string content, Func<TestWorkspace, Task> body, bool outOfProcess)
-        {
-            await TestAsync(content, body, outOfProcess, null);
-            await TestAsync(content, body, outOfProcess, w => new FirstDocIsActiveDocumentTrackingService(w));
-            await TestAsync(content, body, outOfProcess, w => new FirstDocIsVisibleDocumentTrackingService(w));
-            await TestAsync(content, body, outOfProcess, w => new FirstDocIsActiveAndVisibleDocumentTrackingService(w));
+            var exportProvider = ExportProviderCache.GetOrCreateExportProviderFactory(catalog).CreateExportProvider();
```

Comment:
:memo: `GetOrCreateExportProviderFactory` costs about 250ms on a fast machine and returns an immutable object. Given the high rate of reuse of each catalog in the Navigate To tests, this would benefit substantially from caching.

Reply:
shoudl that catching just be inside GetOrCreateExportProviderFactory itself?  or would that be valuable in this code instead?

Reply:
➡️ Implemented in a0a34cc